README.md
rcitoid
=========

[![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![cran checks](https://cranchecks.info/badges/worst/rcitoid)](https://cranchecks.info/pkgs/rcitoid)
[![Build Status](https://travis-ci.com/ropenscilabs/rcitoid.svg?branch=master)](https://travis-ci.com/ropenscilabs/rcitoid)
[![Build status](https://ci.appveyor.com/api/projects/status/yk8vpcdr1rmi7byy?svg=true)](https://ci.appveyor.com/project/sckott/rcitoid)
[![codecov](https://codecov.io/gh/ropenscilabs/rcitoid/branch/master/graph/badge.svg)](https://codecov.io/gh/ropenscilabs/rcitoid)
[![rstudio mirror downloads](http://cranlogs.r-pkg.org/badges/rcitoid)](https://github.com/metacran/cranlogs.app)
[![cran version](https://www.r-pkg.org/badges/version/rcitoid)](https://cran.r-project.org/package=rcitoid)




Client for the Citoid service <https://www.mediawiki.org/wiki/Citoid>

docs: <https://en.wikipedia.org/api/rest_v1/#!/Citation/getCitation>

There are two functions, both of which do the same things, except:

- `cit_oid()`: parses text
- `cit_oid_()`: does not parse text, you can parse later yourself

Even with `cit_oid()` though, you get a list of lists, and you may
want to parse it to a data.frame. See an example below.

## Install

Stable version


```r
install.packages("rcitoid")
```

Development version


```r
devtools::install_github("ropenscilabs/rcitoid")
```

Load the package



```r
library("rcitoid")
```

## get citation data

use underscore method to get text


```r
cit_oid_("10.1108/jd-12-2013-0166")
#> [[1]]
#> [1] "[{\"key\":\"T4RDAMPU\",\"version\":0,\"itemType\":\"journalArticle\",\"tags\":[],\"publicationTitle\":\"Journal of Documentation\",\"volume\":\"71\",\"issue\":\"2\",\"language\":\"en\",\"ISSN\":[\"0022-0418\"],\"date\":\"2015-03-09\",\"pages\":\"253–277\",\"DOI\":\"10.1108/JD-12-2013-0166\",\"url\":\"http://www.emeraldinsight.com/doi/10.1108/JD-12-2013-0166\",\"title\":\"Setting our bibliographic references free: towards open citation data\",\"libraryCatalog\":\"Crossref\",\"accessDate\":\"2019-02-12\",\"shortTitle\":\"Setting our bibliographic references free\",\"author\":[[\"Silvio\",\"Peroni\"],[\"Alexander\",\"Dutton\"],[\"Tanya\",\"Gray\"],[\"David\",\"Shotton\"]],\"source\":[\"Zotero\"]}]"
#> attr(,"type")
#> [1] "json"
```

## get citation data

DOI


```r
cit_oid("10.1108/jd-12-2013-0166")
#> [[1]]
#> [[1]]$key
#> [1] "QJPWEUYT"
#> 
#> [[1]]$version
#> [1] 0
#> 
#> [[1]]$itemType
#> [1] "journalArticle"
#> 
...
```

PMID


```r
cit_oid(30446726)
#> [[1]]
#> [[1]]$key
#> [1] "VRT3CWW3"
#> 
#> [[1]]$version
#> [1] 0
#> 
#> [[1]]$itemType
#> [1] "journalArticle"
#> 
...
```

PMCID


```r
cit_oid("PMC4679344")
#> [[1]]
#> [[1]]$key
#> [1] "VWJ6RJNE"
#> 
#> [[1]]$version
#> [1] 0
#> 
#> [[1]]$itemType
#> [1] "journalArticle"
#> 
...
```

ISBN


```r
cit_oid(1439895619)
#> [[1]]
#> [[1]]$itemType
#> [1] "book"
#> 
#> [[1]]$title
#> [1] "Agroecology : the ecology of sustainable food systems"
#> 
#> [[1]]$oclc
#> [1] "744303838"
#> 
...
```

## parse to data.frame

because the resulting data is nested and can have missing data slots,
it's probably easier to get raw text and manipulate from there.


```r
library(dplyr)

pmid <- c(30446726, 30722046, 30687373, 30688010)
pmcid <- c("PMC4679344", "PMC6347797", "PMC6347793")
isbn <- 1439895619
dois <- c("10.1109/jsac.2011.110806", "10.1007/s00422-006-0078-4",
  "10.5040/9781474219624-0044", "10.1109/icemi.2009.5274826",
  "10.1109/wispnet.2017.8299996")
res <- cit_oid_(id = c(pmid, pmcid, isbn, dois))
tbl_df(bind_rows(lapply(res, jsonlite::fromJSON)))
#> # A tibble: 13 x 35
#>    key   version itemType tags  title ISSN  journalAbbrevia…
#>    <chr>   <int> <chr>    <lis> <chr> <lis> <chr>           
#>  1 X42G…       0 journal… <lis… Enha… <chr… Mucosal Immunol 
#>  2 LNTZ…       0 journal… <lis… Shar… <chr… Mol. Biol. Evol.
#>  3 ADUB…       0 journal… <dat… Resp… <chr… Front Plant Sci 
#>  4 GDTL…       0 webpage  <lis… A mi… <NUL… <NA>            
#>  5 EBZS…       0 journal… <lis… ESMO… <NUL… <NA>            
#>  6 M4GB…       0 journal… <lis… Effi… <NUL… <NA>            
#>  7 M3LT…       0 journal… <lis… Iden… <chr… J Hematol Oncol 
#>  8 <NA>       NA book     <NUL… Agro… <NUL… <NA>            
#>  9 BS3M…       0 journal… <lis… Anti… <chr… <NA>            
#> 10 PNNM…       0 journal… <lis… The … <chr… <NA>            
#> 11 WKX7…       0 book     <lis… The … <NUL… <NA>            
#> 12 47LX…       0 journal… <lis… Desi… <NUL… <NA>            
#> 13 QU6I…       0 confere… <lis… Traf… <NUL… <NA>            
#> # … with 28 more variables: publicationTitle <chr>, date <chr>,
#> #   abstractNote <chr>, DOI <chr>, extra <chr>, libraryCatalog <chr>,
#> #   url <chr>, accessDate <chr>, author <list>, PMID <chr>, source <list>,
#> #   pages <chr>, volume <chr>, shortTitle <chr>, PMCID <chr>,
#> #   language <chr>, websiteTitle <chr>, issue <chr>, oclc <chr>,
#> #   ISBN <list>, edition <chr>, place <chr>, numPages <chr>,
#> #   contributor <list>, publisher <chr>, editor <list>,
#> #   proceedingsTitle <chr>, conferenceName <chr>
```

## Meta

* Please [report any issues or bugs](https://github.com/ropenscilabs/rcitoid/issues)
* License: MIT
* Get citation information for `rcitoid` in R doing `citation(package = 'rcitoid')`
* Please note that this project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.


[![ropensci_footer](https://ropensci.org/public_images/github_footer.png)](https://ropensci.org)


NEWS.md
rcitoid 0.1.0
=============

### NEW FEATURES

* Released to CRAN


tests/testthat/test-cit_oid.R
doi1 <- "10.1108/jd-12-2013-0166"
pmid1 <- 30446726
pmcid1 <- "PMC4679344"
isbn1 <- 1439895619

context("cit_oid: DOI")
test_that("cit_oid DOI", {
  x <- cit_oid(doi1)

  expect_is(x, "list")
  expect_is(x[[1]]$key, "character")
  expect_is(x[[1]]$url, "character")
  expect_equal(x[[1]]$DOI, doi1)
})


context("cit_oid: PMID")
test_that("cit_oid PMID", {
  x <- cit_oid(pmid1)

  expect_is(x, "list")
  expect_is(x[[1]]$key, "character")
  expect_is(x[[1]]$url, "character")
  expect_equal(x[[1]]$DOI, "10.1038/s41385-018-0109-1")
  expect_equal(x[[1]]$PMID, as.character(pmid1))
})


context("cit_oid: ISBN")
test_that("cit_oid ISBN", {
  x <- cit_oid(isbn1)

  expect_is(x, "list")
  expect_null(x[[1]]$key)
  expect_is(x[[1]]$url, "character")
  expect_null(x[[1]]$DOI)
  expect_null(x[[1]]$PMID)
  expect_is(x[[1]]$ISBN, "list")
  expect_equal(length(x[[1]]$ISBN), 2)
})


context("cit_oid: PMCID")
test_that("cit_oid PMCID", {
  x <- cit_oid(pmcid1)

  expect_is(x, "list")
  expect_is(x[[1]]$key, "character")
  expect_is(x[[1]]$url, "character")
  expect_equal(x[[1]]$DOI, "10.1097/IGC.0000000000000609")
  expect_equal(x[[1]]$PMID, "26645990")
  expect_equal(x[[1]]$PMCID, pmcid1)
  expect_is(x[[1]]$author, "list")
  expect_is(x[[1]]$author[[1]], "list")
  expect_is(unlist(x[[1]]$author[[1]]), "character")
})


context("cit_oid: many ids")
test_that("cit_oid many ids", {
  x <- cit_oid(id = c(pmid1, pmcid1, isbn1))

  expect_is(x, "list")
  expect_equal(length(x), 3)
  expect_named(x, NULL)
  expect_is(x[[1]], "list")
  expect_is(x[[2]], "list")
  expect_is(x[[3]], "list")
})



test_that("cit_oid fails well", {
  expect_error(cit_oid(), "argument \"id\" is missing")
  expect_error(cit_oid(mtcars), "id must be of class")
})


tests/testthat/test-cit__oid.R
doi1 <- "10.1108/jd-12-2013-0166"
pmid1 <- 30446726
pmcid1 <- "PMC4679344"
isbn1 <- 1439895619

context("cit_oid_: DOI")
test_that("cit_oid_ DOI", {
  x <- cit_oid_(doi1)

  expect_is(x, "list")
  expect_is(x[[1]], "character")
  expect_is(attr(x[[1]], "type"), "character")
  expect_equal(attr(x[[1]], "type"), "json")
  expect_match(x[[1]], doi1)
})


context("cit_oid_: PMID")
test_that("cit_oid_ PMID", {
  x <- cit_oid_(pmid1)

  expect_is(x, "list")
  expect_is(x[[1]], "character")
  expect_is(attr(x[[1]], "type"), "character")
  expect_equal(attr(x[[1]], "type"), "json")
  expect_match(x[[1]], "10.1038/s41385-018-0109-1")
  expect_match(x[[1]], as.character(pmid1))
})


context("cit_oid_: ISBN")
test_that("cit_oid_ ISBN", {
  x <- cit_oid_(isbn1)

  expect_is(x, "list")
  expect_is(x[[1]], "character")
  expect_is(attr(x[[1]], "type"), "character")
  expect_equal(attr(x[[1]], "type"), "json")
  expect_true(grepl(isbn1, x[[1]]))
})


context("cit_oid_: PMCID")
test_that("cit_oid_ PMCID", {
  x <- cit_oid_(pmcid1)

  expect_is(x, "list")
  expect_is(x[[1]], "character")
  expect_is(attr(x[[1]], "type"), "character")
  expect_equal(attr(x[[1]], "type"), "json")

  expect_match(x[[1]], "10.1097/IGC.0000000000000609")
  expect_match(x[[1]], "26645990")
  expect_match(x[[1]], pmcid1)
})


context("cit_oid_: many ids")
test_that("cit_oid_ many ids", {
  x <- cit_oid_(id = c(pmid1, pmcid1, isbn1))
  
  expect_equal(length(x), 3)
  expect_is(x, "list")
  expect_is(x[[1]], "character")
  expect_is(x[[2]], "character")
  expect_is(x[[3]], "character")
  expect_named(x, NULL)
})

test_that("cit_oid_ fails well", {
  expect_error(cit_oid_(), "argument \"id\" is missing")
  expect_error(cit_oid_(mtcars), "id must be of class")
})


R/http.R
oid_base <- "https://en.wikipedia.org"
oid_path <- "api/rest_v1/data/citation"

# errs <- function(x) {
#   if (x$status_code > 201) {
#     fun <- fauxpas::find_error_class(x$status_code)$new()
#     fun$do_verbose(x)
#   }
# }

# oid_GET <- function(format, id, accept_language, ...) {
#   cli <- crul::HttpClient$new(
#     oid_base,
#     headers = cp(list(
#       "User-Agent" = oid_ua(),
#       "X-USER-AGENT" = oid_ua(),
#       "Accept-Language" = accept_language
#     )),
#     opts = list(...)
#   )
#   paths <- file.path(oid_path, format, id)
#   res <- lapply(paths, function(z) cli$get(z))
#   invisible(lapply(res, errs))
#   return(res)
# }

oid_headers <- function(accept_language) {
  cp(list(
    "User-Agent" = oid_ua(),
    "X-USER-AGENT" = oid_ua(),
    "Accept-Language" = accept_language
  ))
}

errs_asnc <- function(x) {
  if (x > 201) {
    fun <- fauxpas::find_error_class(x)$new()
    fun$do_verbose(x)
  }
}

oid_GET_async <- function(format, id, accept_language, ...) {
  paths <- file.path(oid_path, format, id)
  reqs <- lapply(paths, function(w) {
    crul::HttpRequest$new(
      url = oid_base,
      opts = list(...),
      headers = oid_headers(accept_language)
    )$get(path = w)
  })
  cli <- crul::AsyncVaried$new(.list = reqs)
  cli$request()
  invisible(lapply(cli$status_code(), errs_asnc))
  cli$responses()
}

oid_ua <- function() {
  versions <- c(
    paste0("r-curl/", utils::packageVersion("curl")),
    paste0("crul/", utils::packageVersion("crul")),
    sprintf("rOpenSci(rcitoid/%s)",
            utils::packageVersion("rcitoid"))
  )
  paste0(versions, collapse = " ")
}


R/cit_oid.R
#' Get Citoid data
#'
#' @export
#' @param id (character) id of an article, DOI, ISBN, PMCID or PMID
#' @param format (character) the format to use for the resulting citation data.
#' one of mediawiki (default), zotero, mediawiki-basefields, bibtex
#' @param accept_language (character) for some articles the result depends on
#' the accept language value, so provide it if localized content is required
#' @param ... curl options passed on to [crul::verb-GET]
#' @details `cit_oid_()` gets raw text (either bibtex or JSON), and `cit_oid()`
#' parses the text as appropriate for the type
#' @return list of lists or character, see
#' http://opencitations.net/index/coci/api/v1 for explanation of the resulting
#' columns
#' @references https://en.wikipedia.org/api/rest_v1/#!/Citation/getCitation,
#' https://www.mediawiki.org/wiki/Citoid
#' @examples 
#' url<-"https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/30446726"
#' if (crul::ok(url)) {
#'   pmid1 <- 30446726
#'   cit_oid(pmid1)
#' }
#' 
#' \dontrun{
#' doi1 <- "10.1108/jd-12-2013-0166"
#' doi2 <- "10.1371/journal.pone.0058568"
#' pmid1 <- 30446726
#' pmcid1 <- "PMC4679344"
#' isbn1 <- 1439895619
#'
#' # different formats
#' cit_oid(doi1)
#' cit_oid(pmid1, format = "mediawiki")
#' cit_oid(pmid1, format = "zotero")
#' cit_oid(pmid1, format = "mediawiki-basefields")
#' cat(cit_oid(pmid1, format = "bibtex")[[1]])
#'
#' # PMID example
#' cit_oid(pmid1, verbose = TRUE)
#'
#' # ISBN example
#' cit_oid(isbn1, verbose = TRUE)
#'
#' # PMCID example
#' cit_oid(pmcid1)
#'
#' # set the accept language
#' x <- cit_oid(pmid1, accept_language = "fr-FR", verbose = TRUE)
#' x <- cit_oid(doi2, accept_language = "de-DE", verbose = TRUE)
#'
#' # just get raw text/json
#' cit_oid_(pmcid1)
#'
#' # many ids at once
#' cit_oid(id = c(pmid1, pmcid1, isbn1))
#' cit_oid_(id = c(pmid1, pmcid1, isbn1))
#' cit_oid_(id = c(pmid1, pmcid1, isbn1), format = "bibtex")
#' }
cit_oid <- function(id, format = "mediawiki", accept_language = NULL, ...) {
  x <- cit_oid_(id, format, accept_language, ...)
  lapply(x, function(w) {
    switch(
      attr(w, "type"),
      bibtex = w,
      json = jsonlite::fromJSON(w, FALSE)[[1]]
    )
  })
}

#' @export
#' @rdname cit_oid
cit_oid_ <- function(id, format = "mediawiki", accept_language = NULL, ...) {
  assert(id, c("character", "numeric", "integer"))
  assert(format, "character")
  assert(accept_language, "character")

  if (!format %in% oid_formats) {
    stop("'format' must be one of: ", paste0(oid_formats, collapse = ", "))
  }
  x <- oid_GET_async(format, curl::curl_escape(id), accept_language, ...)
  lapply(x, function(w) structure(w$parse("UTF-8"), type = get_ctype(w)))
}

oid_formats <- c("mediawiki", "zotero", "mediawiki-basefields", "bibtex")

get_ctype <- function(x) {
  ctype <- x$response_headers$`content-type`
  if (grepl("bibtex", ctype)) {
    "bibtex"
  } else if (grepl("json", ctype)) {
    "json"
  }
}
